### ISSUE TITLE
[BUG]: Incorrect filtering of BPF assembler

### ISSUE THREAD

**Issue by yonghong-song** [compiler-explorer/compiler-explorer#5701]
### Describe the bug

The short link for the program: https://godbolt.org/z/rb9P1sTfo

For the following program:
```
static unsigned long x; // new atomics 64-bit only.
static unsigned long y;

#define READ_ONCE(a) (*(volatile typeof(a) *)&a)
#define WRITE_ONCE(a, v) ((*(volatile typeof(a) *)&a) = v)

int foo(void)
{ 
    unsigned long var = 0;

    var = READ_ONCE(x);
    if (var) {
        WRITE_ONCE(y, 1);
    } else {
        WRITE_ONCE(y, 42);
    }
    return var;
}
```
With 'BPF gcc thunk':
```
foo:
        lddw    %r0,x
        lddw    %r1,y
        ldxdw   %r0,[%r0+0]
        jeq     %r0,0,.L2
        stdw    [%r1+0],1
        exit
.L2:
        stdw    [%r1+0],42
        exit
x:
        .zero   8
```
It is correct.
With my local trunk clang compiler (built from latest llvm-project main branch),
```
$ clang --target=bpf -O2 -c t.c && llvm-objdump -d --no-show-raw-insn t.o

t.o:    file format elf64-bpf

Disassembly of section .text:

0000000000000000 <foo>:
       0:       r1 = 0x0 ll
       2:       r0 = *(u64 *)(r1 + 0x0)
       3:       r1 = 0x2a
       4:       if r0 == 0x0 goto +0x1 <LBB0_2>
       5:       r1 = 0x1

0000000000000030 <LBB0_2>:
       6:       r2 = 0x8 ll
       8:       *(u64 *)(r2 + 0x0) = r1
       9:       exit
```
The generated code slightly different from gcc's. But still correct.

With compiler explorer, the BPF clang trunk generates:
```
foo:                                    # @foo
        r0 = *(u64 *)(r1 + 0)
        if r0 == 0 goto LBB0_2
LBB0_2:
        *(u64 *)(r2 + 0) = r1
        exit
```
The code is not correct. The 'r1'  and 'r2' are not even defined.

The same thing happens with x86_64 clang trunk compiler with '-O2 -target bpf' option.
But if with x86_64 clang trunk compiler with '-O2', the generated code is correct.

I suspect the clang trunk compiler might be built having some issues with BPF target.
Could you share how clang trunk compiler is built including 'revision', 'cmake flags' etc. so we can 
reproduce the issue and make recommendation from BPF side whether anything might not be correct.

### Steps to reproduce

1. Goto https://godbolt.org/z/rb9P1sTfo
2. Select 'C', 'BPF clang (Trunk)' and '-O2' option
3. Check the result

### Expected behavior

Idealy the result will be similar to below:
```
0000000000000000 <foo>:
       0:       r1 = 0x0 ll
       2:       r0 = *(u64 *)(r1 + 0x0)
       3:       r1 = 0x2a
       4:       if r0 == 0x0 goto +0x1 <LBB0_2>
       5:       r1 = 0x1

0000000000000030 <LBB0_2>:
       6:       r2 = 0x8 ll
       8:       *(u64 *)(r2 + 0x0) = r1
       9:       exit
```
which is the result in my local build. But the asm code does not need to have precise match
as long as the final asm code is correct.

### Reproduction link

https://godbolt.org/z/rb9P1sTfo

### Screenshots

Not applicable

**Comment by yonghong-song** [comment]
cc @4ast @paulmckrcu

**Comment by dkm** [comment]
Everything is available in a dedicated project. See the build script: https://github.com/compiler-explorer/clang-builder/blob/main/build/build.sh
Everything runs in docker, which config file is also in the same project: https://github.com/compiler-explorer/clang-builder/blob/main/Dockerfile
It's built nightly and you can check what really happens here: https://github.com/compiler-explorer/compiler-workflows/actions/workflows//build-daily-clang.yml


You can also fetch the compiler tarball if you want to experiment locally with the compiler we are using.

```
$ cd infra
$ make ce
$ ./bin/ce_install   --enable nightly  install clang|grep trunk
```
Does that help with your investigations?

**Comment by yonghong-song** [comment]
Thanks @dkm, I will check the above links to do some investigation.

**Comment by OfekShilon** [comment]
@yonghong-song :
It seems like a compiler-explorer internal issue. As a workaround, uncheck the 'Filters.. / Unused labels' checkbox and the full code will be displayed:
![image](https://github.com/compiler-explorer/compiler-explorer/assets/73080/31ba78a4-b690-4efc-a3e9-a4122b86ad25)

**Comment by yonghong-song** [comment]
@OfekShilon Thanks. Indeed with the above workaround, the displayed asm code looks correct.

**Comment by dkm** [comment]
Of course it's the assembler filters again :). It should learn a bit about the latest C-ish BPF syntax

**Comment by OfekShilon** [comment]
This seems to be fixed by #5706. @yonghong-song can you please verify and close?

**Comment by yonghong-song** [comment]
> This seems to be fixed by #5706. @yonghong-song can you please verify and close?

Yes, it does fix the problem. Will close the issue.
